src/representation_theory/fixed_space.lean

```diff
@@ -0,0 +1,110 @@
+/-
+Copyright (c) 2022 Antoine Labelle. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Labelle
+-/
+import representation_theory.basic
+
+/-!
+# Fixed space of a group representation
+
+This file introduce the subspace of fixed points of a group representation
+and proves basic result about it.
+The main tool used is the average of all elements of the group, seen as an element of
+`monoid_algebra k G`. Scalar multiplication by this special element gives a projection onto the
+subspace of fixed vectors.
+In order for the definition of the average element to make sense, we need to assume for most of the
+results that the order of `G` is invertible in `k` (e. g. `k` has characteristic `0`).
+-/
+
+open_locale big_operators
+open monoid_algebra
+open finset
+open finite_dimensional
+open linear_map
+open representation
+
+section average
+variables (k G : Type*) [comm_semiring k] [group G]
+variables [fintype G] [invertible (fintype.card G : k)]
+
+/--
+The average of all elements of the group `G`, considered as an element of `monoid_algebra k G`.
+-/
+noncomputable def average : monoid_algebra k G :=
+  ⅟(fintype.card G : k) • ∑ g : G, of k G g
+
+lemma average_def : average k G = ⅟(fintype.card G : k) • ∑ g : G, of k G g := rfl
+
+/--
+`average k G` is invariant under left multiplication by elements of `G`.
+-/
+@[simp]
+theorem mul_average_left (g : G) :
+  (finsupp.single g 1 * average k G : monoid_algebra k G) = average k G :=
+begin
+  simp [average_def, finset.mul_sum],
+  set f : G → monoid_algebra k G := λ x, finsupp.single x 1,
+  show ⅟ ↑(fintype.card G) • ∑ (x : G), f (g * x) = ⅟ ↑(fintype.card G) • ∑ (x : G), f x,
+  rw function.bijective.sum_comp (group.mul_left_bijective g) _,
+end
+
+/--
+`average k G` is invariant under right multiplication by elements of `G`.
+-/
+@[simp]
+theorem mul_average_right (g : G) :
+  average k G * finsupp.single g 1 = average k G :=
+begin
+  simp [average_def, finset.sum_mul],
+  set f : G → monoid_algebra k G := λ x, finsupp.single x 1,
+  show ⅟ ↑(fintype.card G) • ∑ (x : G), f (x * g) = ⅟ ↑(fintype.card G) • ∑ (x : G), f x,
+  rw function.bijective.sum_comp (group.mul_right_bijective g) _,
+end
+
+end average
+
+section fixed_space
+variables (k G V : Type*) [comm_semiring k] [group G] [add_comm_group V]
+variables [module k V] [distrib_mul_action G V] [smul_comm_class G k V]
+
+/--
+The subspace of vectors fixed by all elements of `G`
+-/
+def fixed_space : submodule k V :=
```

Comment:
Should we upgrade this to a submodule over `monoid_algebra k G`?

Reply:
I guess we could, though I'm not sure it would be particularly helpful since the action is trivial

Reply:
It's not the action that's interesting, it's having this as a point in the lattice `submodule (monoid_algebra k G)`.

Reply:
In this case, should we do both a submodule over `k` and over `monoid_algebra k G`? Because I think it's also useful to have it as a point in the lattice of sub-`k`-vector spaces.

Reply:
Sure, but we can get that as `submodule.restrict_scalars` of the other one.

Reply:
It's done, though I also keeped the definition as a `submodule k V` because the easiest way I could find to define a submodule over `monoid_algebra k G` is from a `submodule k V` which is stable under multiplication by elements of `G`

Reply:
Isn't `invariants` a more common name? And there's also coinvariants.

Reply:
Maybe? Both make sense I think but yeah, I can change the name to `invariants` if you believe that's more common.

Reply:
Let's wait for more people to weigh in :)

Reply:
opened a poll on Zulip 5 hours ago and the score is 5:0 for `invariants` as of now: https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/naming.20contest/near/275771137

Reply:
Well that seems pretty unanimous, I'll change the name!